crates/bevy_ecs/src/query/direct.rs

```diff
@@ -0,0 +1,336 @@
+use crate::{
+    component::Component,
+    entity::Entity,
+    query::{
+        Fetch, FilterFetch, QueryEntityError, QueryIter, QueryState, ReadOnlyFetch, WorldQuery,
+    },
+    system::{QueryComponentError, QuerySingleError},
+    world::{Mut, World},
+};
+use bevy_tasks::TaskPool;
+use std::any::TypeId;
+
+/// [DirectQuery] is a wrapper for [QueryState] that owns it. It is used for one-time direct queries
+/// on the [World].
+pub struct DirectQuery<'w, Q: WorldQuery, F: WorldQuery = ()>
```

Comment:
We already discussed this elsewhere, but I'm not sold on this abstraction because it encourages users to not use query cacheing (which can have big perf implications). It feels like a band-aid fix for the severe reduction in QueryState usability.

Reply:
I agree. How do we restore the QueryState usability?

Reply:
Following your recommendation, I removed `DirectQuery` and only made `QueryState::*manual()` take the counters explicitly. Most of the API changes have been reverted thanks to that. Fixed in 7d38cc4426abe99143cc479bf5befb63de8020c0

Reply:
Awesome thats what i had in mind. Can you push it / merge it here (along with the other changes)

Reply:
I unfortunately cannot push to the PR, I'm relying on @alice-i-cecile to do it.

Reply:
This has been resolved; we remembered to just give @Davier write permissions to my fork.